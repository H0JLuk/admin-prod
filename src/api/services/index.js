import { getSession } from './sessionService';

export const getReqOptions = () => {
    const token = getSession();
    return {
        headers: { token }
    };
};
