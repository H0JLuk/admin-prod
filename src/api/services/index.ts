import { ApiRequestHeaders, ApiRequestOptions } from '@types';
import { getDefaultAppCode } from './clientAppService';
import { getAppCode, getSession } from './sessionService';

const TOKEN_HEADER = 'token';
const APP_CODE_HEADER = 'clientAppCode';
const CONTENT_TYPE = 'Content-Type';
const ACCEPT = 'Accept';

export const getReqOptions = (contentType?: string, currAppCode?: string): ApiRequestOptions => {
    const token = getSession();
    const appCode = currAppCode || getAppCode() || getDefaultAppCode(); // TODO: remove getDefaultAppCode after second phase
    const headers = {} as ApiRequestHeaders;
    token && (headers[TOKEN_HEADER] = token);
    appCode && (headers[APP_CODE_HEADER] = appCode);
    contentType && (headers[ACCEPT] = '*/*');
    contentType && (headers[CONTENT_TYPE] = contentType);
    return { headers };
};
