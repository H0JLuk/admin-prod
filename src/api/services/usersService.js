import {Api} from '../apiClient';
import { getReqOptions } from './index';

export async function addUser(data) {
    return Api.post('/admin/user', data, getReqOptions);
}