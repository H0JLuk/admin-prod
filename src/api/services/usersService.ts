import { Api } from '../apiClient';
import { getReqOptions } from './index';
import {
    DefaultApiResponse,
    RegisterUserRequest,
    ResetUserPassword,
    SaveUserResponse,
    UpdateUserRequest,
    UpdateUsersSalePoint,
    UserDto,
    UserInfo,
    UserPaginationResponse,
} from '@types';

export function getUser(pn: string) {
    return Api.get<UserInfo>(`/admin/user/${pn}`, getReqOptions());
}

export function oldAddUser(data: UserDto) {
    return Api.post<SaveUserResponse>('/admin/user', data, getReqOptions());
}

export function addUser(data: RegisterUserRequest) {
    return Api.post<DefaultApiResponse & SaveUserResponse>('/admin/user/register', data, getReqOptions());
}

export function oldRemoveUser(pn: string) {
    return Api.delete<DefaultApiResponse>(`/admin/user/${pn}`, getReqOptions());
}

export function removeUser(userId: number) {
    return Api.delete<DefaultApiResponse>(`/admin/user/delete/${userId}`, getReqOptions());
}

export function getUsersList(searchParams: string) {
    return Api.get<UserPaginationResponse>(`/admin/user/filtered?${searchParams}`, getReqOptions());
}

// TODO: раскомментировать после добавления на бэке запроса на удаление нескольких пользователей
/*
export function removeUsers(usersIds: number[]) {
    return Api.delete(`/admin/user/${usersIds}`, getReqOptions());
}
*/

export function resetUser(pn: string) {
    return Api.post<ResetUserPassword>(`/admin/user/reset/${pn}`, {}, getReqOptions());
}

export function unblockUser(pn: string) {
    return Api.post<DefaultApiResponse>(`/admin/user/unblock/${pn}`, {}, getReqOptions());
}

export function saveUser(id: number, data: UpdateUserRequest) {
    return Api.post<DefaultApiResponse & { newPassword?: string; }>(`/admin/user/edit/${id}`, data, getReqOptions());
}

export function editLocationAndSalePointUsers(data: UpdateUsersSalePoint) {
    return Api.post<DefaultApiResponse>('/admin/user/editSalePoint', data, getReqOptions());
}

function getReqForFileUpload(appCode: string) {
    const options = getReqOptions('multipart/form-data');
    options.headers.clientAppCode = appCode;
    return options;
}

export function addUsersWithTemplate(appCode: string, data: FormData) {
    const options = getReqForFileUpload(appCode);
    return Api.upload<Blob>('/admin/user/upload/file', data, options, 'blob');
}

export function deleteUsersWithTemplate(appCode: string, data: FormData) {
    const options = getReqForFileUpload(appCode);
    return Api.upload<Blob>('/admin/user/delete/file', data, options, 'blob');
}
