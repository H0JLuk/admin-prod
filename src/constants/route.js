export const ROUTE = {
    CORE:           '/',
    APP:            '/app',
    LOGIN:          '/login',
    AUDITOR:        '/auditor',
    ADMIN:          '/admin',
    OWNER:          '/owner',
    USER_MANAGER:   '/user-manager',
    CLIENT_APPS:    '/client-apps',
    OLD_DESIGN:     '/old_design',
};

const PAGE = {
    AUDIT:          'audit',
    CATEGORY:       'category',
    DZO:            'dzo',
    DASHBOARD:      'dashboard',
    FILES:          'files',
    PRESENTATION:   'presentation',
    PROMO_CAMPAIGN: 'promo-campaign',
    USERS:          'users',
    APPS:           'client-apps',
};

export const USERS_PAGES = {
    ADD_USER:        '/create',
    USER_INFO:       '/:userId',
    EDIT_USER:       '/:userId/edit',
    EDIT_SOME_USERS: '/multiple_edit',
};

export const PROMO_CAMPAIGN_PAGES = {
    ADD_PROMO_CAMPAIGN:        '/create',
    PROMO_CAMPAIGN_INFO:       '/:promoCampaignId/info',
    PROMO_CAMPAIGN_EDIT:       '/:promoCampaignId/edit',
    VISIBILITY_SETTINGS:       '/:promoCampaignId/visibility-setting',
    LOAD_PROMO_CODES:          '/:promoCampaignId/load-promo-codes',
    PROMO_CODES_STATISTIC:     '/:promoCampaignId/promo-codes-statistic',
};

export const CLIENT_APPS_PAGES = {
    ADD_APP:        '/create-app',
    EDIT_APP:       '/app-properties',
};

export const DZO_PAGES = {
    ADD_DZO:        '/create',
    DZO_INFO:       '/:dzoId/info',
    DZO_EDIT:       '/:dzoId/edit',
};

export const ROUTE_AUDITOR = {
    AUDIT:          `${ ROUTE.AUDITOR }/${ PAGE.AUDIT }`
};

export const ROUTE_USER_MANAGER = {
    USERS:          `${ ROUTE.USER_MANAGER }/${ PAGE.USERS }`,
};

export const ROUTE_ADMIN = {
    OLD_DESIGN_USERS: `${ ROUTE.OLD_DESIGN }${ ROUTE.ADMIN }/${ PAGE.USERS }`,
    OLD_DESIGN_PROMO_CAMPAIGN: `${ ROUTE.OLD_DESIGN }${ ROUTE.ADMIN }/${ PAGE.PROMO_CAMPAIGN }`,
    DASHBOARD:      `${ ROUTE.ADMIN }/${ PAGE.DASHBOARD }`,
    USERS:          `${ ROUTE.ADMIN }/${ PAGE.USERS }`,
    FILES:          `${ ROUTE.ADMIN }/${ PAGE.FILES }`,
    DZO:            `${ ROUTE.ADMIN }/${ PAGE.DZO }`,
    PRESENTATION:   `${ ROUTE.ADMIN }/${ PAGE.PRESENTATION }`,
    CATEGORY:       `${ ROUTE.ADMIN }/${ PAGE.CATEGORY }`,
    PROMO_CAMPAIGN: `${ ROUTE.ADMIN }/${ PAGE.PROMO_CAMPAIGN }`,
    OLD_DESIGN_APPS: `${ ROUTE.OLD_DESIGN }${ ROUTE.ADMIN }/${ PAGE.APPS }`,
    APPS:           `${ ROUTE.ADMIN }/${ PAGE.APPS }`,
    APP_SETTINGS:   `${ ROUTE.ADMIN }/${ PAGE.APPS }${ CLIENT_APPS_PAGES.EDIT_APP }`,
};

export const ROUTE_OWNER = {
    OLD_DESIGN_PROMO_CAMPAIGN: `${ ROUTE.OLD_DESIGN }${ ROUTE.OWNER }/${ PAGE.PROMO_CAMPAIGN }`,
    DASHBOARD:      `${ ROUTE.OWNER }/${ PAGE.DASHBOARD }`,
    DZO:            `${ ROUTE.OWNER }/${ PAGE.DZO }`,
    PRESENTATION:   `${ ROUTE.OWNER }/${ PAGE.PRESENTATION }`,
    FILES:          `${ ROUTE.OWNER }/${ PAGE.FILES }`,
    CATEGORY:       `${ ROUTE.OWNER }/${ PAGE.CATEGORY }`,
    PROMO_CAMPAIGN: `${ ROUTE.OWNER }/${ PAGE.PROMO_CAMPAIGN }`,
    OLD_DESIGN_APPS: `${ ROUTE.OLD_DESIGN }${ ROUTE.OWNER }/${ PAGE.APPS }`,
    APPS:           `${ ROUTE.OWNER }/${ PAGE.APPS }`,
};
