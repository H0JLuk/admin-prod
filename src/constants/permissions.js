import { getRole } from '../api/services/sessionService';
import ROLES from './roles';

const DEFAULT_USERS_INTERACTIONS = {
    editUser: [],
    resetUserPassword: [],
    unlockUser: [],
    deleteUser: [],
    canSelectUserInTable: [],
};

const USERS_INTERACTIONS_BY_ROLE = {
    [ROLES.ADMIN]: {
        editUser: [ROLES.USER],
        resetUserPassword: [ROLES.USER],
        unlockUser: [ROLES.USER],
        deleteUser: [ROLES.USER],
        canSelectUserInTable: [ROLES.USER],
    },
    [ROLES.USER_MANAGER]: {
        editUser: [ROLES.USER],
        resetUserPassword: [ROLES.USER],
        unlockUser: [ROLES.USER],
        deleteUser: [ROLES.USER],
        canSelectUserInTable: [ROLES.USER],
    },
    [ROLES.AUDITOR]: {
        ...DEFAULT_USERS_INTERACTIONS,
    },
    [ROLES.PRODUCT_OWNER]: {
        ...DEFAULT_USERS_INTERACTIONS,
    },
};

/**
 * Объект, который определяет какие взаимодействия доступны для ролей в отношении пользователей с другими ролями
 * На выходе получится объект вида:
 * {
 *    Admin: {
 *       Admin: {
 *          editUser: boolean,
 *          resetUserPassword: boolean,
 *          unlockUser: boolean,
 *          deleteUser: boolean,
 *       },
 *       User: {....}
 *    }
 * }
 * Чтобы получить объект с настройками взаимодействия текущего пользователя
 * с другим выбранным пользователем нужно первым ключём передать роль текущего пользователя,
 * а вторым параметром передать роль выбранного пользователя
 *
 * INTERACTIONS_CURR_USER_WITH_OTHER_USER[getRole()][userRole]; // где `userRole` роль выбранного пользователя
 */
export const INTERACTIONS_CURR_USER_WITH_OTHER_USER = Object.values(ROLES).reduce((allInteractions, currRole) => {
    const currInteractions = USERS_INTERACTIONS_BY_ROLE[currRole] || DEFAULT_USERS_INTERACTIONS;

    const interactionsForOtherUser = Object.values(ROLES).reduce((interactionsForOtherRole, otherRole) => {
        const interactionsCurrUserWithOtherUser = Object.entries(currInteractions).reduce((result, [key, value]) => ({
            ...result,
            [key]: value.includes(otherRole),
        }), {});

        return {
            ...interactionsForOtherRole,
            [otherRole]: interactionsCurrUserWithOtherUser,
        };
    }, {});

    return {
        ...allInteractions,
        [currRole]: {
            ...interactionsForOtherUser,
        },
    };
}, {});

export function getCurrUserInteractions() {
    return INTERACTIONS_CURR_USER_WITH_OTHER_USER[getRole()] || {};
}

export function getCurrUserInteractionsForOtherUser(role) {
    return getCurrUserInteractions()[role] || {};
}
