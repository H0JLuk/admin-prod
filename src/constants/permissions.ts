import { getRole } from '../api/services/sessionService';
import ROLES from './roles';

export type UsersInteractions = Record<string, ROLES[]>;

const DEFAULT_USERS_INTERACTIONS: UsersInteractions = {
    editUser: [],
    resetUserPassword: [],
    unlockUser: [],
    deleteUser: [],
    canSelectUserInTable: [],
    canGenerateQRCode: [],
};

const USERS_INTERACTIONS_BY_ROLE: Record<ROLES, UsersInteractions> = {
    [ROLES.ADMIN]: {
        editUser: [ROLES.USER, ROLES.PARTNER, ROLES.EXTERNAL_USER, ROLES.REFERAL_LINK],
        resetUserPassword: [ROLES.USER, ROLES.PARTNER, ROLES.EXTERNAL_USER, ROLES.REFERAL_LINK],
        unlockUser: [ROLES.USER, ROLES.PARTNER, ROLES.EXTERNAL_USER, ROLES.REFERAL_LINK],
        deleteUser: [ROLES.USER, ROLES.PARTNER, ROLES.EXTERNAL_USER, ROLES.REFERAL_LINK],
        canSelectUserInTable: [ROLES.USER, ROLES.PARTNER, ROLES.EXTERNAL_USER, ROLES.REFERAL_LINK],
        canGenerateQRCode: [ROLES.USER, ROLES.PARTNER, ROLES.EXTERNAL_USER, ROLES.REFERAL_LINK],
    },
    [ROLES.USER_MANAGER]: {
        editUser: [ROLES.USER],
        resetUserPassword: [ROLES.USER],
        unlockUser: [ROLES.USER],
        deleteUser: [ROLES.USER],
        canSelectUserInTable: [ROLES.USER],
    },
    [ROLES.AUDITOR]: {
        ...DEFAULT_USERS_INTERACTIONS,
    },
    [ROLES.PRODUCT_OWNER]: {
        ...DEFAULT_USERS_INTERACTIONS,
    },
    [ROLES.USER]: {
        ...DEFAULT_USERS_INTERACTIONS,
    },
    [ROLES.PARTNER]: {
        editUser: [ROLES.EXTERNAL_USER, ROLES.REFERAL_LINK],
        resetUserPassword: [ROLES.EXTERNAL_USER, ROLES.REFERAL_LINK],
        unlockUser: [ROLES.EXTERNAL_USER, ROLES.REFERAL_LINK],
        deleteUser: [ROLES.EXTERNAL_USER, ROLES.REFERAL_LINK],
        canSelectUserInTable: [ROLES.EXTERNAL_USER, ROLES.REFERAL_LINK],
    },
    [ROLES.EXTERNAL_USER]: {
        ...DEFAULT_USERS_INTERACTIONS,
    },
    [ROLES.REFERAL_LINK]: {
        ...DEFAULT_USERS_INTERACTIONS,
    },
};

export type CommonPermissions = {
    canSetUserRole: boolean;
    canSetUserPartner: boolean;
    canSetLoginType: boolean;
};

const COMMON_PERMISSIONS_BY_ROLE: Record<ROLES, CommonPermissions> = {
    [ROLES.ADMIN]: {
        canSetUserRole: true,
        canSetUserPartner: true,
        canSetLoginType: true,
    },
    [ROLES.USER_MANAGER]: {
        canSetUserRole: false,
        canSetUserPartner: false,
        canSetLoginType: true,
    },
    [ROLES.AUDITOR]: {
        canSetUserRole: false,
        canSetUserPartner: false,
        canSetLoginType: false,
    },
    [ROLES.PRODUCT_OWNER]: {
        canSetUserRole: false,
        canSetUserPartner: false,
        canSetLoginType: false,
    },
    [ROLES.USER]: {
        canSetUserRole: false,
        canSetUserPartner: false,
        canSetLoginType: false,
    },
    [ROLES.PARTNER]: {
        canSetUserRole: false,
        canSetUserPartner: false,
        canSetLoginType: false,
    },
    [ROLES.EXTERNAL_USER]: {
        canSetUserRole: false,
        canSetUserPartner: false,
        canSetLoginType: false,
    },
    [ROLES.REFERAL_LINK]: {
        canSetUserRole: false,
        canSetUserPartner: false,
        canSetLoginType: false,
    },
};

/**
 * Объект, который определяет какие взаимодействия доступны для ролей в отношении пользователей с другими ролями
 * На выходе получится объект вида:
 * {
 *    Admin: {
 *       Admin: {
 *          editUser: boolean,
 *          resetUserPassword: boolean,
 *          unlockUser: boolean,
 *          deleteUser: boolean,
 *       },
 *       User: {....}
 *    }
 * }
 * Чтобы получить объект с настройками взаимодействия текущего пользователя
 * с другим выбранным пользователем нужно первым ключём передать роль текущего пользователя,
 * а вторым параметром передать роль выбранного пользователя
 *
 * INTERACTIONS_CURR_USER_WITH_OTHER_USER[getRole()][userRole]; // где `userRole` роль выбранного пользователя
 */

export type InteractionsCurrUserToOtherUser = Record<string, boolean>;
export type InteractionsForOtherUser = Record<ROLES, InteractionsCurrUserToOtherUser>;
export type InteractionsCurrUserWIthOtherUser = Record<ROLES, InteractionsForOtherUser>;

export const INTERACTIONS_CURR_USER_WITH_OTHER_USER = Object.values(ROLES).reduce((allInteractions, currRole) => {
    const currInteractions = USERS_INTERACTIONS_BY_ROLE[currRole] || DEFAULT_USERS_INTERACTIONS;

    const interactionsForOtherUser = Object.values(ROLES).reduce((interactionsForOtherRole, otherRole) => {
        const interactionsCurrUserWithOtherUser = Object.entries(currInteractions).reduce((result, [key, value]) => ({
            ...result,
            [key]: value.includes(otherRole),
        }), {} as InteractionsCurrUserToOtherUser);

        return {
            ...interactionsForOtherRole,
            [otherRole]: interactionsCurrUserWithOtherUser,
        };
    }, {} as InteractionsForOtherUser);

    return {
        ...allInteractions,
        [currRole]: {
            ...interactionsForOtherUser,
        },
    };
}, {} as InteractionsCurrUserWIthOtherUser);

export function getCommonPermissionsByRole() {
    return COMMON_PERMISSIONS_BY_ROLE[getRole()] || {};
}

export function getCurrUserInteractions() {
    return INTERACTIONS_CURR_USER_WITH_OTHER_USER[getRole()] || {};
}

export function getCurrUserInteractionsForOtherUser(role: ROLES) {
    return getCurrUserInteractions()[role] || {};
}
