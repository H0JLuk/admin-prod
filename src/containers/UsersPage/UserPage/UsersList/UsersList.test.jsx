import React from 'react';
import { generatePath } from 'react-router-dom';
import UserList from './UsersList';
import { fireEvent, render, act, screen } from '@testing-library/react';
import { sleep } from '../../../../setupTests';
import { permissionsTestData, usersListTestData, usersTestArray } from '../../../../../__tests__/constants';
import * as usersService from '../../../../api/services/usersService';
import { shallow } from 'enzyme';
import { USERS_PAGES } from '../../../../constants/route';
import * as permissions from '../../../../constants/permissions';

const DEFAULT_URL_SEARCH_PARAMS = 'pageNo=0&pageSize=10&sortBy=&direction=ASC&filterText=';
const diffLoginTypeUsersListTestData = {
    ...usersListTestData,
    users: [
        ...usersListTestData['users'].slice(10),
        { ...usersListTestData['users'][11], loginType: 'SBER_REGISTRY', id: 22882 }
    ],
};

jest.mock('../../../../components/TableDeleteModal/TableDeleteModal', () => ({ modalClose, visible, refreshTable }) => (
    <div>
        { visible && (
            <div>
                <div>Modal</div>
                <button onClick={ modalClose }>close</button>
                <button onClick={ refreshTable }>refresh</button>
            </div>
        ) }
    </div>
));
jest.mock('./EmptyUsersPage/EmptyUsersPage', () => () => (<div>EmptyUsersPage</div>));
jest.mock('../UserForm', () => () => (<div>UserForm</div>));

const mockPush = jest.fn();

jest.mock('../../../../utils/notifications', () => ({
    customNotifications: {
        success: jest.fn(),
        error: jest.fn(),
    }
}));

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useHistory: () => ({
        push: mockPush,
        replace: jest.fn(),
        location: {
            search: '?pageNo=0&pageSize=10&sortBy=&direction=ASC&filterText=',
        },
    }),
    useLocation: () => ({
        search: '?pageNo=0&pageSize=10&sortBy=&direction=ASC&filterText=',
    }),
    generatePath: jest.fn(),
}));

beforeEach(() => {
    usersService.getUsersList = jest.fn().mockResolvedValue(usersListTestData);
    permissions.getCurrUserInteractions = jest.fn().mockReturnValue(permissionsTestData);
});

describe('<UsersList /> test', () => {
    const props = {
        matchPath: '/admin/users',
    };

    it('UserList snapshot', () => {
        const wrapper = shallow(<UserList { ...props } />);

        expect(wrapper.debug()).toMatchSnapshot();
    });

    it('should run getUsersList fn with default search params', async () => {
        await act(async () => {
            render(<UserList { ...props } />);
            await sleep();

            expect(usersService.getUsersList).toBeCalledWith(DEFAULT_URL_SEARCH_PARAMS);
        });
    });

    it('should show loading message while fetching', async () => {
        await act(async () => {
            const { getByText } = render(<UserList { ...props } />);

            expect(getByText(/Загрузка.../)).toBeInTheDocument();
        });
    });

    it('should load list of user after fetch', async () => {
        await act(async () => {
            const { getByText } = render(<UserList { ...props } />);
            await sleep();

            expect(getByText(/Краснозаводск/)).toBeInTheDocument();
        });
    });

    it('should show error message rejected from loadUsersData', async () => {
        const spy = jest.spyOn(console, 'error').mockImplementation(() => '');
        usersService.getUsersList = jest.fn().mockRejectedValue('error');
        await act(async () => {
            render(<UserList { ...props } />);
            await sleep();

            expect(spy).toBeCalled();
        });
    });

    it('should run history.push on add user btn click', async () => {
        await act(async () => {
            const { getAllByText } = render(<UserList { ...props } />);
            await sleep();
            const addBtn = getAllByText(/Добавить/)[0];

            fireEvent.click(addBtn);
            expect(mockPush).toBeCalledWith(`${props.matchPath}${USERS_PAGES.ADD_USER}`);
        });
    });

    it('should switch chooseBtn to cancelBtn on click & vice versa', async () => {
        await act(async () => {
            const { getByText } = render(<UserList { ...props } />);
            await sleep();

            fireEvent.click(getByText(/Выбрать/));
            expect(getByText(/Отменить/)).toBeInTheDocument();
            expect(getByText(/Выбрать все/)).toBeInTheDocument();

            fireEvent.click(getByText(/Отменить/));
            expect(getByText(/Выбрать/)).toBeInTheDocument();
        });
    });

    it('should switch chooseAllBtn to cancelSelectionBtn on click & vice versa', async () => {
        await act(async () => {
            const { getByText } = render(<UserList { ...props } />);
            await sleep();

            fireEvent.click(getByText(/Выбрать/));
            fireEvent.click(getByText(/Выбрать все/));
            fireEvent.click(getByText(/Отменить выбор/));

            expect(getByText(/Выбрать все/)).toBeInTheDocument();
        });
    });

    it('should fire onclick fn for table row', async () => {
        render(<UserList { ...props } />);
        await sleep();

        fireEvent.click(document.querySelector('.ant-table-row'));
        expect(generatePath).toBeCalledWith(`${props.matchPath}${USERS_PAGES.USER_INFO}`, { userId: 1 });
        expect(mockPush).toBeCalled();
    });

    it('should redirect on edit btn click when items selected', async () => {
        await act(async () => {
            const { getByText } = render(<UserList { ...props } />);
            await sleep();

            fireEvent.click(getByText(/Выбрать/));
            fireEvent.click(document.querySelectorAll('.ant-checkbox-input')[2]);
            fireEvent.click(getByText(/Редактировать/));
            expect(mockPush).toBeCalledWith(`${props.matchPath}${USERS_PAGES.EDIT_SOME_USERS}`, { users: [usersTestArray[2]] });
        });
    });

    it('should redirect on row click', async () => {
        generatePath.mockReturnValue(`${props.matchPath}/4`);
        await act(async () => {
            const { container } = render(<UserList { ...props } />);
            await sleep();

            fireEvent.click(container.querySelector('.ant-table-row'));
            expect(mockPush).toBeCalled();
            expect(mockPush).toBeCalledWith(`${props.matchPath}/4`);
        });
    });

    it('should select/unselect on row click', async () => {
        await act(async () => {
            const { container, getByText } = render(<UserList { ...props } />);
            await sleep();

            fireEvent.click(getByText(/Выбрать/));
            const tr = container.querySelectorAll('.ant-table-row')[2];

            fireEvent.click(tr);
            expect(getByText(/Выбрано 1/)).toBeInTheDocument();

            fireEvent.click(tr);
            expect(getByText(/Выбрано 0/)).toBeInTheDocument();
        });
    });

    it('should not select on row click', async () => {
        await act(async () => {
            const { container, getByText } = render(<UserList { ...props } />);
            await sleep();

            fireEvent.click(getByText(/Выбрать/));
            fireEvent.click(container.querySelector('.ant-table-row'));
            expect(getByText(/Выбрано 0/)).toBeInTheDocument();
        });
    });

    it('should reset pass', async () => {
        usersService.resetUser = jest.fn().mockResolvedValue({});
        await act(async () => {
            const { getByText } = render(<UserList { ...props } />);
            await sleep();

            fireEvent.click(getByText(/Выбрать/));
            fireEvent.click(getByText(/Выбрать все/));
            fireEvent.click(getByText(/Сбросить пароль/));
            await sleep();

            expect(usersService.resetUser).toBeCalled();
        });
    });

    it('should not reset pass according to diff login type test data', async () => {
        usersService.resetUser = jest.fn().mockResolvedValue({});
        usersService.getUsersList = jest.fn().mockResolvedValue(diffLoginTypeUsersListTestData);
        await act(async () => {
            const { getByText } = render(<UserList { ...props } />);
            await sleep();

            fireEvent.click(getByText(/Выбрать/));
            fireEvent.click(getByText(/Выбрать все/));
            fireEvent.click(getByText(/Сбросить пароль/));
            await sleep();

            expect(usersService.resetUser).not.toBeCalled();
        });
    });

    it('should successfully resets passwords with rejected one too', async () => {
        usersService.resetUser = jest.fn().mockRejectedValueOnce('error').mockResolvedValue({});
        await act(async () => {
            render(<UserList { ...props } />);
        });
        await sleep();

        await act(async () => {
            fireEvent.click(screen.getByText('Выбрать'));
        });

        await act(async () => {
            fireEvent.click(screen.getByText('Выбрать все'));
        });

        await act(async () => {
            fireEvent.click(screen.getByText('Сбросить пароль'));
        });

        expect(screen.getByText('Выбрано 0')).toBeInTheDocument();
    });

    it('should reject Promise.all while resetting passwords', async () => {
        Promise.allSettled = jest.fn().mockRejectedValueOnce([ { status: 'error' } ]);
        const spy = jest.spyOn(console, 'warn').mockImplementation(() => '');
        await act(async () => {
            render(<UserList { ...props } />);
        });
        await sleep();

        await act(async () => {
            fireEvent.click(screen.getByText('Выбрать'));
        });

        await act(async () => {
            fireEvent.click(screen.getByText('Выбрать все'));
        });

        await act(async () => {
            fireEvent.click(screen.getByText('Сбросить пароль'));
        });

        expect(spy).toBeCalled();
    });

    it('should call `getUsersList` when changed page', async () => {
        await act(async () => {
            const { container } = render(<UserList { ...props } />);
            await sleep();

            fireEvent.click(container.querySelector('.ant-pagination-item.ant-pagination-item-2'));
            await sleep();

            expect(usersService.getUsersList).toBeCalled();
        });
    });

    it('should toggle modal', async () => {
        await act(async () => {
            const { getByText, queryAllByText } = render(<UserList { ...props } />);
            await sleep();

            fireEvent.click(getByText(/Выбрать/));
            fireEvent.click(getByText(/Выбрать все/));
            fireEvent.click(getByText(/Удалить/));
            expect(getByText(/Modal/)).toBeInTheDocument();

            fireEvent.click(getByText(/close/));
            expect(queryAllByText(/Modal/)).toHaveLength(0);
        });
    });

    it('should run refreshTable ', async () => {
        await act(async () => {
            const { getByText } = render(<UserList { ...props } />);
            await sleep();

            fireEvent.click(getByText(/Выбрать/));
            fireEvent.click(getByText(/Выбрать все/));
            fireEvent.click(getByText(/Удалить/));
            fireEvent.click(getByText(/refresh/));

            expect(usersService.getUsersList).toBeCalledTimes(2);
        });
    });
});
