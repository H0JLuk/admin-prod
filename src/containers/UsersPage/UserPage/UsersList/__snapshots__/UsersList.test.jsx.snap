// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UsersList /> test UserList snapshot 1`] = `
"<div className=\\"loadingPage\\">
  Загрузка...
</div>"
`;
